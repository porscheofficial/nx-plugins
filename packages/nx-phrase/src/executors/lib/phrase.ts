import FormData from "form-data"
import { RequestError, requestUrl } from "./request"
import debug from "debug"

const logger = debug("nx-plugins.nx-phrase.lib.phrase")

const BASE_URL = process.env.PHRASE_API_BASE_URL ?? "https://api.phrase.com/v2"

export type Translations = Record<string, string>

export interface PhraseLocale {
    id: string
    name: string
    code: string
    default: boolean
    main: boolean
    rtl: boolean
    plural_forms: string[]
    created_at: string
    updated_at: string
    source_locale: {
        id: string
        name: string
        code: string
    }
    fallback_locale: {
        id: string
        name: string
        code: string
    }
}

export interface PhraseKey {
    id: string
    name: string
    description: string
    name_hash: string
    plural: boolean
    max_characters_allowed: number
    tags: string[]
    created_at: string
    updated_at: string
}

export interface PhraseClientConfig {
    token: string
}

type PhraseAllPaginated<T> = Omit<T, "page" | "per_page">

export class PhraseClient {
    #baseUrl = BASE_URL
    private ITEMS_PER_PAGE = 20
    private credentials: { token: string }

    constructor(config: PhraseClientConfig) {
        this.credentials = { token: config.token }

        logger(`Using ${this.#baseUrl} as base`)
    }

    public async localesList({
        projectId,
        ...optionalArgs
    }: {
        projectId: string
        // These arguments are snake cased because that is how the api expects them
        page?: number
        per_page?: number
        sort_by?: string
        branch?: string
    }): Promise<PhraseLocale[]> {
        const url = new URL(`${this.#baseUrl}/projects/${projectId}/locales`)
        for (const argName in optionalArgs) {
            if (optionalArgs[argName]) {
                url.searchParams.set(argName, optionalArgs[argName])
            }
        }

        const response = await requestUrl(url, {
            headers: {
                Authorization: `token ${this.credentials.token}`,
            },
            successStatus: 200,
            errorMessagePrefix: "Failed to load locales",
        })

        return JSON.parse(response) as Promise<PhraseLocale[]>
    }

    public async localesListAll(args: PhraseAllPaginated<Parameters<PhraseClient["localesList"]>[0]>) {
        let page = 0
        let items = [] as PhraseLocale[]
        let nextPage = [] as PhraseLocale[]

        do {
            nextPage = await this.localesList({ ...args, page, per_page: this.ITEMS_PER_PAGE })
            items = [...items, ...nextPage]

            page++
        } while (nextPage.length >= this.ITEMS_PER_PAGE)

        return items
    }

    public async localesDownload({
        projectId: projectId,
        id,
        ...otherArgs
    }: {
        projectId: string
        id: string
        // These arguments are snake cased because that is how the api expects them
        file_format: string
        branch?: string
    }): Promise<string> {
        logger("Downloading translation for", id)

        const url = new URL(`${this.#baseUrl}/projects/${projectId}/locales/${id}/download`)
        for (const argName in otherArgs) {
            if (otherArgs[argName]) {
                url.searchParams.set(argName, otherArgs[argName])
            }
        }

        const downloadRequestFactory = () =>
            requestUrl(url, {
                headers: {
                    Authorization: `token ${this.credentials.token}`,
                },
                successStatus: 200,
                errorMessagePrefix: `Failed to download translations for ${id}`,
            })

        return this.retry(downloadRequestFactory, (e: unknown, attempt: number) => {
            if (e instanceof RequestError) {
                // handle concurrency limit exceeded
                if (e.statusCode === 429 && attempt < 5) {
                    logger(`Retrying due to concurrency`)
                    return true
                }
            }

            return false
        })
    }

    private async retry<T extends () => R | Promise<R>, R>(
        callbackFn: T,
        shouldRetry: (error: unknown, attempt: number) => boolean | Promise<boolean>
    ): Promise<R> {
        let attempt = 0
        let repeat = false
        do {
            try {
                attempt++
                return await callbackFn()
            } catch (e) {
                repeat = await shouldRetry(e, attempt)
                if (!repeat) {
                    throw e
                }

                const sleepFor = 2 ** attempt * 1000
                await this.sleepHelper(sleepFor)
            }
        } while (repeat)

        throw new Error("Unreachable")
    }

    /**
     * Sleep helper can be overwritten with jest.mock to allow proper testing  of retry
     */
    async sleepHelper(time: number) {
        await new Promise((resolve) => setTimeout(resolve, time))
    }

    public async upload(
        {
            projectId,
            ...otherArgs
        }: {
            projectId: string
            // These arguments are snake cased because that is how the api expects them
            locale_id: string
            update_translations?: boolean
            skip_upload_tags?: boolean
            file_format: string
            branch?: string
        },
        fileContent: string | Buffer
    ) {
        const url = new URL(`${this.#baseUrl}/projects/${projectId}/uploads`)
        const body = new FormData()
        for (const argName in otherArgs) {
            if (otherArgs[argName]) {
                body.append(argName, `${otherArgs[argName]}`)
            }
        }
        if (Buffer.isBuffer(fileContent)) {
            body.append("file", fileContent.toString(), { filename: "default.json" })
        } else {
            body.append("file", fileContent, { filename: "default.json" })
        }

        await requestUrl(url, {
            headers: {
                Authorization: `token ${this.credentials.token}`,
            },
            method: "POST",
            requestBody: body,
            successStatus: 201,
            errorMessagePrefix: "Failed to upload default translations",
        })
    }

    public async resolveKeyIdByName({
        projectId,
        keyName,
        ...otherArgs
    }: {
        projectId: string
        keyName: string
        branch?: string
    }) {
        const url = new URL(`${this.#baseUrl}/projects/${projectId}/keys/search`)
        url.searchParams.set("q", `name:${keyName}`)
        for (const argName in otherArgs) {
            if (otherArgs[argName]) {
                url.searchParams.set(argName, otherArgs[argName])
            }
        }

        const response = await requestUrl(url, {
            method: "POST",
            headers: {
                Authorization: `token ${this.credentials.token}`,
                Accept: "application/json",
                "Content-Type": "application/json",
            },
            successStatus: 200,
            errorMessagePrefix: "Failed to load locales",
        })

        const foundKeys = JSON.parse(response) as PhraseKey[]
        return foundKeys.map((key) => key.id)
    }

    public async deleteKey({ projectId, keyId, ...otherArgs }: { projectId: string; keyId: string; branch?: string }) {
        const url = new URL(`${this.#baseUrl}/projects/${projectId}/keys/${keyId}`)
        for (const argName in otherArgs) {
            if (otherArgs[argName]) {
                url.searchParams.set(argName, otherArgs[argName])
            }
        }

        await requestUrl(url, {
            method: "DELETE",
            headers: {
                Authorization: `token ${this.credentials.token}`,
            },
            successStatus: 204,
            errorMessagePrefix: "Failed to delete key",
        })
    }
}
