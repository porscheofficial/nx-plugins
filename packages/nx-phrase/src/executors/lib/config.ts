import { existsSync, readFileSync } from "fs-extra"
import { resolve } from "path"

import { NonSensitiveArgs } from "./types"
import { ExecutorContext } from "@nrwl/devkit"
import { load } from "js-yaml"
import { PhraseClientConfig } from "./phrase"
import { BuildExecutorSchema } from "../build/schema"

interface ConfigFileFormat {
    phrase: {
        [appName: string]: {
            access_token: string
        }
    }
}

export interface InternalPhraseConfig extends NonSensitiveArgs {
    phraseClientConfig: PhraseClientConfig
}

function readConfig(context: ExecutorContext) {
    const projectName = context.projectName
    if (!projectName) {
        throw new Error("project name not set in context")
    }

    const configFileLocation = resolve(context.root, ".phrase.yml")

    if (!existsSync(configFileLocation)) {
        console.error(`Could not find configuration file at ${configFileLocation}`)
        process.exit(-1)
    }

    const configFileContent = load(readFileSync(configFileLocation, { encoding: "utf8" })) as ConfigFileFormat

    let error = false
    if (!configFileContent.phrase) {
        console.info(`Config file does not have 'phrase' root node.`)
        error = true
    }
    const configSection = configFileContent.phrase[projectName]
    if (!configSection) {
        console.info(`Config file has no configuration section for project '${projectName}'`)
        error = true
    }

    if (!configSection.access_token) {
        console.error(
            `'access_token' not set in .phrase.yml for translation target in project.json of project '${projectName}'.`
        )
        error = true
    }

    if (error) {
        process.exit(-1)
    }

    return configSection
}

function validateConfig(config: NonSensitiveArgs, projectName: string, requiredConfigProperties: string[]) {
    let error = false

    requiredConfigProperties
        .filter((requiredProperty) => !config[requiredProperty])
        .forEach((requiredProperty) => {
            console.info(
                `Configuration property '${requiredProperty}' not set for translation target in project.json of project '${projectName}'.`
            )
            error = true
        })

    return !error
}

export function getConfig(
    options: BuildExecutorSchema,
    context: ExecutorContext,
    requiredConfigurationProperties: string[]
): InternalPhraseConfig {
    if (!context.projectName) {
        throw new Error("project name not set in context")
    }

    const {
        projectName,
        workspace: { projects },
    } = context
    const { sourceRoot, root } = projects[projectName]

    const rawConfig = readConfig(context)
    // get output first, so it does not get overwritten
    const output = resolve(
        root,
        // use from project.json -> config file -> fallback
        options.output ?? "./translations"
    )
    const config = {
        output,
        branch: options.branch,
        fileFormat: options.fileFormat ?? "react_simple_json",
        ignoreGlob: options.ignoreGlob ?? "**/*.d.ts",
        projectId: options.projectId,
        sourceRoot: options.sourceRoot ?? sourceRoot,
        sourceGlob: options.sourceGlob ?? "**/*.{ts,tsx}",
        uploadLanguageId: options.uploadLanguageId || "default",
        sourceKeyTransformer: options.sourceKeyTransformer,
        phraseKeyTransformer: options.phraseKeyTransformer,
        sourceKeyFilter: options.sourceKeyFilter,
        phraseKeyFilter: options.phraseKeyFilter,
        useFallbackLocale: options.useFallbackLocale,
        useSourceLocaleAsFallback: options.useSourceLocaleAsFallback,
    } as NonSensitiveArgs

    if (!validateConfig(config, projectName, requiredConfigurationProperties)) {
        throw new Error("Invalid configuration")
    }

    const phraseClientConfig: PhraseClientConfig = { token: rawConfig.access_token }

    return {
        phraseClientConfig,
        ...config,
    }
}
