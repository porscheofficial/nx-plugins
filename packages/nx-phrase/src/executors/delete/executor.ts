import { ExecutorContext } from "@nrwl/devkit"
import { existsSync } from "fs"

import { getConfig, InternalPhraseConfig } from "../lib/config"
import { deleteKeys } from "../lib/delete"
import { NonSensitiveArgs } from "../lib/types"

const requiredConfigProperties = ["projectId", "inputFile"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, requiredConfigProperties)
    const { projectName } = context

    // get path from config / cli param?
    if (!existsSync(options.inputFile)) {
        throw new Error(`input file "${options.inputFile}" does not exist.`)
    }

    // load file
    await deleteKeys(options.inputFile, config)

    return { success: true }
}
