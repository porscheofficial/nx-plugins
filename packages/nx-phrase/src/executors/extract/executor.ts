import { ExecutorContext } from "@nx/devkit"

import { getConfig, InternalPhraseConfig } from "../../lib/config"
import { NonSensitiveArgs } from "../../lib/types"
import { extract } from "../../lib/extract"

export const pullRequiredConfigs = ["output"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, pullRequiredConfigs)

    await extract(config, context)

    console.log("Done.")

    return { success: true }
}
