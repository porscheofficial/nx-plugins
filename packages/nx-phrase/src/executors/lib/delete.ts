import { readFileSync } from "fs"
import { InternalPhraseConfig } from "./config"
import { PhraseClient } from "./phrase"

export async function deleteKeys(file: string, config: InternalPhraseConfig): Promise<void> {
    const keysToDelete = JSON.parse(readFileSync(file).toString())

    // ask for confirmation?
    const phrase = new PhraseClient(config.phraseClientConfig)
    for (const key of keysToDelete) {
        // resolve keyIds for key name
        const keyIds = await phrase.resolveKeyIdByName({
            projectId: config.projectId,
            keyName: key,
            branch: config.branch,
        })
        // delete keyIds
        for (const keyId of keyIds) {
            try {
                await phrase.deleteKey({ projectId: config.projectId, keyId, branch: config.branch })
            } catch (e) {
                console.log(`Could not delete key with id '${keyId}' for name '${key}', ${e}`)
            }
        }
    }
}
