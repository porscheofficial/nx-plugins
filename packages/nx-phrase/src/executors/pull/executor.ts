import { ExecutorContext } from "@nrwl/devkit"

import { getConfig, InternalPhraseConfig } from "../lib/config"
import { pull } from "../lib/pull"
import { NonSensitiveArgs } from "../lib/types"

export const pullRequiredConfigs = ["projectId", "output"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, pullRequiredConfigs)

    await pull(config)

    console.log("Done.")

    return { success: true }
}
