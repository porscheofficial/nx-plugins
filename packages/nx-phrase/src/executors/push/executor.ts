import { ExecutorContext } from "@nx/devkit"

import { getConfig, InternalPhraseConfig } from "../../lib/config"
import { push } from "../../lib/push"
import { NonSensitiveArgs } from "../../lib/types"

export const pushRequiredConfigs = ["projectId", "uploadLanguageId"]

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, pushRequiredConfigs)

    await push(config, context)

    console.log("Done.")

    return { success: true }
}
