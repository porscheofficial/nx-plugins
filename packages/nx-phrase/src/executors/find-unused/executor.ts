import { ExecutorContext } from "@nrwl/devkit"
import { existsSync, writeFileSync } from "fs"
import { resolve } from "path"
import { prepareOutput } from "../../utils"

import { getConfig, InternalPhraseConfig } from "../lib/config"
import { PullHelper } from "../lib/pull"
import { extractTranslations } from "../lib/push"
import { NonSensitiveArgs } from "../lib/types"

const requiredConfigProperties = ["projectId", "output"]

function defaultTransformKeyFn(translationKey: string) {
    return translationKey
}
type transformTranslationKeyFn = typeof defaultTransformKeyFn

function defaultFilterKeyFn(translationKey: string) {
    return !!translationKey
}
type filterKeyFn = typeof defaultFilterKeyFn

async function loadTransformer<T = transformTranslationKeyFn>(path: string) {
    if (path) {
        if (existsSync(path)) {
            const filterFunction = (await import(path)).default as T
            if (typeof filterFunction === "function") {
                return filterFunction
            } else {
                throw new Error(`${path} does not export a function as default`)
            }
        } else {
            throw new Error(`${path} does not exist`)
        }
    }
}

async function getKeysFromSource(config: InternalPhraseConfig, outputFilePath: string) {
    // extracts current translations
    const compiledTranslationsFilePath = await extractTranslations(config, outputFilePath)
    const compiledTranslationKeys = Object.keys((await import(compiledTranslationsFilePath)).default)

    return compiledTranslationKeys
}

async function getKeysFromPhrase(config: InternalPhraseConfig) {
    const pull = new PullHelper({ ...config, fileFormat: "react_simple_json" })

    // downloads available translations
    const locales = await pull.listLocales()
    const localeToRawTranslations = await pull.downloadTranslations(locales)
    const keysInPhrase = new Set<string>()
    Object.keys(localeToRawTranslations).forEach((locale) =>
        Object.keys(JSON.parse(localeToRawTranslations[locale])).forEach((key) => keysInPhrase.add(key))
    )

    // use set for deduplication
    const keysInPhraseList = [...keysInPhrase]

    return keysInPhraseList
}

function applyTransform(input: string[], transformer: transformTranslationKeyFn) {
    // apply optional filter (use set for deduplication)
    const set = new Set<string>()
    input.map(transformer).forEach((key) => set.add(key))
    return [...set]
}

function filterList(input: string[], filter: filterKeyFn) {
    return input.filter(filter)
}

export default async function runExecutor(options: Partial<NonSensitiveArgs>, context: ExecutorContext) {
    const config: InternalPhraseConfig = getConfig(options, context, requiredConfigProperties)
    const { projectName } = context

    // extract and prepare keys from source code
    const outputPath = await prepareOutput({
        projectRoot: context.workspace.projects[context?.projectName]?.root ?? context.root,
        subfolder: "unused",
        workingDirectory: options.workingDirectory,
    })
    let sourceTranslationKeys
    {
        const sourceKeyFilter = options.sourceKeyFilter
            ? await loadTransformer<filterKeyFn>(resolve(context.root, options.sourceKeyFilter))
            : defaultFilterKeyFn

        const sourceKeyTransformer = options.sourceKeyTransformer
            ? await loadTransformer(resolve(context.root, options.sourceKeyTransformer))
            : defaultTransformKeyFn

        const allSourceTranslationKeys = await getKeysFromSource(config, outputPath)
        sourceTranslationKeys = filterList(allSourceTranslationKeys, sourceKeyFilter)

        if (allSourceTranslationKeys.length !== sourceTranslationKeys.length) {
            const bogusKeys = allSourceTranslationKeys.filter((key) => !sourceTranslationKeys.includes(key))
            const sourceBugKeysFilename = `${outputPath}/${projectName}.filtered-source-keys.json`
            writeFileSync(sourceBugKeysFilename, JSON.stringify(bogusKeys, null, 2), {
                flag: "w",
            })
            console.log(`Keys filtered from source written to: ${sourceBugKeysFilename}`)
        }

        sourceTranslationKeys = applyTransform(sourceTranslationKeys, sourceKeyTransformer)
    }

    // extract and prepare keys from phrase
    let phraseTranslationKeys
    {
        const phraseKeyFilter = options.phraseKeyFilter
            ? await loadTransformer<filterKeyFn>(resolve(context.root, options.phraseKeyFilter))
            : defaultFilterKeyFn

        const phraseKeyTransformer = options.phraseKeyTransformer
            ? await loadTransformer(resolve(context.root, options.phraseKeyTransformer))
            : defaultTransformKeyFn

        const allPhraseTranslationKeys = await getKeysFromPhrase(config)
        phraseTranslationKeys = filterList(allPhraseTranslationKeys, phraseKeyFilter)

        if (allPhraseTranslationKeys.length !== phraseTranslationKeys.length) {
            const bogusKeys = allPhraseTranslationKeys.filter((key) => !phraseTranslationKeys.includes(key))
            const phraseBugKeysFilename = `${outputPath}/${projectName}.filtered-phrase-keys.json`
            writeFileSync(phraseBugKeysFilename, JSON.stringify(bogusKeys, null, 2), {
                flag: "w",
            })
            console.log(`Keys filtered from phrase written to: ${phraseBugKeysFilename}`)
        }

        phraseTranslationKeys = applyTransform(phraseTranslationKeys, phraseKeyTransformer)
    }

    // compare available and current translations
    const notUploadedYet = sourceTranslationKeys.filter((key) => !phraseTranslationKeys.includes(key))
    const unusedKeys = phraseTranslationKeys.filter((key) => !sourceTranslationKeys.includes(key))

    // tell difference
    const unusedKeysFilename = `${outputPath}/${projectName}.unused-keys.json`
    writeFileSync(unusedKeysFilename, JSON.stringify(unusedKeys, null, 2), {
        flag: "w",
    })
    console.log(`Unused keys written to: ${unusedKeysFilename}`)

    const pendingUploadKeysFilename = `${outputPath}/${projectName}.pending-upload-keys.json`
    writeFileSync(pendingUploadKeysFilename, JSON.stringify(notUploadedYet, null, 2), {
        flag: "w",
    })
    console.log(`Keys pending upload written to: ${pendingUploadKeysFilename}`)

    // console.log(`${context.targetName} ${context.configurationName}`)

    console.log("Done.")

    return { success: true }
}
