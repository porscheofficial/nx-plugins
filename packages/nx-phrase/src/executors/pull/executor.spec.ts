import { resolve } from "path"

import { ExecutorContext } from "@nx/devkit"
import nock from "nock"

import executor from "./executor"
import { NonSensitiveArgs } from "../../lib/types"
import { PhraseClient } from "../../lib/phrase"
import { readFileSync } from "fs"
import { PhraseFileFormat } from "../../lib/consts"

const options: Partial<NonSensitiveArgs> = {
    projectId: "projectId",
    output: ".nx-phrase/translations",
}

const TEST_ASSETS_DIR = resolve(__dirname, "../../../test")

function nockForProject(projectId = "project_id") {
    const french = { greeting: "bonjour" }

    nock("https://api.phrase.com/v2")
        .get(`/projects/${projectId}/locales`)
        .matchHeader("Authorization", /token .*/)
        .query({ per_page: 20, page: 1 })
        .replyWithFile(200, `${TEST_ASSETS_DIR}/localesListResponse.json`)

        .get(/\/projects\/[^/]+\/locales\/[^/]+\/download/)
        .matchHeader("Authorization", /token .*/)
        .query({ file_format: PhraseFileFormat.REACT_SIMPLE_JSON })
        .thrice()
        .reply(429, { message: "Concurrency limit exceeded" })

        .get(/\/projects\/[^/]+\/locales\/french\/download/)
        .matchHeader("Authorization", /token .*/)
        .query({ file_format: PhraseFileFormat.REACT_SIMPLE_JSON })
        .reply(200, french)

        .get(/\/projects\/[^/]+\/locales\/[^/]+\/download/)
        .matchHeader("Authorization", /token .*/)
        .query({ file_format: PhraseFileFormat.REACT_SIMPLE_JSON })
        .twice()
        .replyWithFile(200, `${TEST_ASSETS_DIR}/localeDownloadResponse.json`)
}

// Remove actual sleep from implementation
jest.spyOn(PhraseClient.prototype, "sleepHelper").mockImplementation(() => Promise.resolve())

describe("Pull", () => {
    const context = {
        root: TEST_ASSETS_DIR,
        projectName: "test_app",
        workspace: { projects: { test_app: { root: TEST_ASSETS_DIR } } } as unknown,
    } as ExecutorContext

    afterEach(() => {
        expect(nock.isDone()).toBe(true)
    })

    it("can pull", async () => {
        nockForProject(options.projectId)
        const output = await executor(options, context)
        expect(output.success).toBe(true)

        expect(
            readFileSync(resolve(TEST_ASSETS_DIR, ".nx-phrase/translations", "fr-BE.json")).toString()
        ).toMatchSnapshot()
        expect(
            readFileSync(resolve(TEST_ASSETS_DIR, ".nx-phrase/translations", "fr-CA.json")).toString()
        ).toMatchSnapshot()
    })

    it("works with fallback_locale workaround", async () => {
        nockForProject(options.projectId)
        const output = await executor({ ...options, useSourceLocaleAsFallback: true }, context)
        expect(output.success).toBe(true)

        // canada has a source_locale set, so it should contain all keys from the source language as well
        const canadianFrench = JSON.parse(
            readFileSync(resolve(TEST_ASSETS_DIR, ".nx-phrase/translations", "fr-CA.json")).toString()
        )
        expect(canadianFrench).toHaveProperty("greeting")

        // belgion does not have source_locale set, so it should NOT contain the key defined in french
        const belgianFrench = JSON.parse(
            readFileSync(resolve(TEST_ASSETS_DIR, ".nx-phrase/translations", "fr-BE.json")).toString()
        )
        expect(belgianFrench).not.toHaveProperty("greeting")
    })
})
