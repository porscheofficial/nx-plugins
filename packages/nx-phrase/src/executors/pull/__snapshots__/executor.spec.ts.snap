// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pull can pull 1`] = `
"{
    \\"a.b.c.d\\": \\"Foo\\",
    \\"a.c.d.e\\": \\"Bar\\",
    \\"a.d.e.c\\": \\"FooBar\\",
    \\"o.l.e.o\\": \\"BarFoo\\",
    \\"K.u.i.l\\": \\"I have spoken\\",
    \\"K.u.i.l[extended]\\": \\"I have spoken\\",
    \\"K.u.i.l[\\": \\"I have spoken\\",
    \\"K.u.i.l]\\": \\"I have spoken\\",
    \\"K.u.i.l[]\\": \\"I have spoken\\",
    \\"K.u.i.l[,a]\\": \\"I have spoken\\",
    \\"K.u.i.l[a,]\\": \\"I have spoken\\",
    \\"K.u.i.l[ ]\\": \\"I have spoken\\",
    \\"K.u.i.l[ , a ]\\": \\"I have spoken\\",
    \\"K.u.i.l[ a , ]\\": \\"I have spoken\\",
    \\"K.u.i.l[b,a]\\": \\"I have spoken\\",
    \\"K.u.i.l[a,b]\\": \\"I have spoken\\"
}
"
`;

exports[`Pull can pull 2`] = `
"{
    \\"a.b.c.d\\": \\"Foo\\",
    \\"a.c.d.e\\": \\"Bar\\",
    \\"a.d.e.c\\": \\"FooBar\\",
    \\"o.l.e.o\\": \\"BarFoo\\",
    \\"K.u.i.l\\": \\"I have spoken\\",
    \\"K.u.i.l[extended]\\": \\"I have spoken\\",
    \\"K.u.i.l[\\": \\"I have spoken\\",
    \\"K.u.i.l]\\": \\"I have spoken\\",
    \\"K.u.i.l[]\\": \\"I have spoken\\",
    \\"K.u.i.l[,a]\\": \\"I have spoken\\",
    \\"K.u.i.l[a,]\\": \\"I have spoken\\",
    \\"K.u.i.l[ ]\\": \\"I have spoken\\",
    \\"K.u.i.l[ , a ]\\": \\"I have spoken\\",
    \\"K.u.i.l[ a , ]\\": \\"I have spoken\\",
    \\"K.u.i.l[b,a]\\": \\"I have spoken\\",
    \\"K.u.i.l[a,b]\\": \\"I have spoken\\"
}
"
`;
